### ISSUE TITLE
path.insertAfter should respect trailingComments

### ISSUE THREAD

**Issue by restrry** [babel/babel#7002]
Choose one: is this a bug report or feature request?
not sure whether that's a bug or not.

### Input Code
```js
function hello() {
    ...
}

//# sourceMappingURL=source.js.map
```

### Plugin code
```js
export default function (babel) {
  const { types: t } = babel;
  
  return {
    visitor: {
	  FunctionDeclaration(path, state){
            const declaration = t.expressionStatement(
                t.assignmentExpression(
                    "=",
                    t.memberExpression(t.identifier('name'), t.identifier('prop')),
                    t.objectExpression([])
                )
            );
            path.insertAfter(declaration);
        }
     }
  };
}
```

### Output Code (Current behaviour)
```js
function hello() {
   ....
}
//# sourceMappingURL=source.js.map
name.prop = {};
```

### Output Code (Expected behaviour)
```js
function hello() {
   ....
}
name.prop = {};
//# sourceMappingURL=source.js.map
```

### Babel/Babylon Configuration (.babelrc, package.json, cli command)
Parser: babylon7-7.0.0-beta.31
Transformer: babelv7-7.0.0-beta.3

### Expected Behavior
`path.insertAfter` should place new node before the trailing comment (probably there is a problem how node-comment are parsed, but modification logic is correct, I'm not sure)

### Current Behavior
parser consider comment as a part of node and  traverser with `insertAfter` put new node behind the comment. that's a problem, especially for source map url comment.

sandbox:
http://astexplorer.net/#/gist/d411201151b805f80352b51478f48110/81b913729bef2deadfee7319c72470d24852baf3

**Comment by nicolo-ribaudo** [comment]
I think we should attach the comments at the start and at the end of the code to the `Program` node, so that their position doesn't change.

**Comment by Andarist** [comment]
@nicolo-ribaudo do you suggest special handling for `Program`'s node comments?

**Comment by loganfsmyth** [comment]
We don't generally guarantee much in the way of comment placement for output. Maybe we could consider handling this case specifically by removing the source map comment from the parsed content entirely, then adding it back on after transformation?

**Comment by xtuc** [comment]
I agree with @loganfsmyth, the source map handling done by Babel should be completely transparent to the plugin author.

Basically: `output + sourcemap`  at code generation.

**Comment by dickeylth** [comment]
@nicolo-ribaudo I'm wondering what's the difference between `FunctionDeclaration` & `ExpressionStatement`, I'm encountering the issue that `insertBefore` won't respect `leadingComments`: <https://github.com/babel/babel/issues/7102#issuecomment-353829115>, and the result in `FunctionDeclaration` demo is just what I expected, however the `ExpressionStatement` would insert code after `leadingComments`.

**Comment by develar** [comment]
Yeah, regression in the babel 7. module convertor adds `exports.AppInfo = AppInfo;` _after_ `//# sourceMappingURL=appInfo.js.map` and as result debugger / sourcemap support is broken.

**Comment by xtuc** [comment]
Sorry @develar what do you mean by "module convertor"?

**Comment by develar** [comment]
@xtuc `@babel/plugin-transform-modules-commonjs` Configuration — https://github.com/develar/ts-babel/blob/master/babel-preset/index.js#L16

Workaround: https://github.com/develar/ts-babel/blob/master/src/builder.ts#L86
